### PR TITLE
generate systemd pod

### PR DESCRIPTION
**Pull request by vrothberg**
generate systemd: support pods

Support generating systemd unit files for a pod.  Podman generates one
unit file for the pod including the PID file for the infra container's
conmon process and one unit file for each container (excluding the infra
container).

Note that this change implies refactorings in the `pkg/systemdgen` API.

The output of a simple pod with an alpine container looks as follows:
```
[Unit]
Description=5d85e91228e02b0fca3982e6db9a0458a381eeffca7951d06f415d6a6495fa29 Podman Pod
Documentation=man:podman-generate-systemd(1)
Requires=container-63202b5664827c0922807cb62ea7ec5c931faabca2c730e5baa29f201bab3596.service
Before=container-63202b5664827c0922807cb62ea7ec5c931faabca2c730e5baa29f201bab3596.service

[Service]
Restart=on-failure
ExecStart=/home/valentin/go/src/github.com/containers/libpod/bin/podman pod start 5d85e91228e02b0fca3982e6db9a0458a381eeffca7951d06f415d6a6495fa29
ExecStop=/home/valentin/go/src/github.com/containers/libpod/bin/podman pod stop  5d85e91228e02b0fca3982e6db9a0458a381eeffca7951d06f415d6a6495fa29
KillMode=none
Type=forking
PIDFile=/tmp/1000/overlay-containers/ae2965240e1ed93064d7145373346d0461c9474215837d364b830bafc7106006/userdata/conmon.pid

[Install]
WantedBy=multi-user.target

[Unit]
Description=63202b5664827c0922807cb62ea7ec5c931faabca2c730e5baa29f201bab3596 Podman Container
Documentation=man:podman-generate-systemd(1)
PartOf=pod-5d85e91228e02b0fca3982e6db9a0458a381eeffca7951d06f415d6a6495fa29.service.service

[Service]
Restart=on-failure
ExecStart=/home/valentin/go/src/github.com/containers/libpod/bin/podman start 63202b5664827c0922807cb62ea7ec5c931faabca2c730e5baa29f201bab3596
ExecStop=/home/valentin/go/src/github.com/containers/libpod/bin/podman -t 10 63202b5664827c0922807cb62ea7ec5c931faabca2c730e5baa29f201bab3596
KillMode=none
Type=forking
PIDFile=/tmp/1000/overlay-containers/63202b5664827c0922807cb62ea7ec5c931faabca2c730e5baa29f201bab3596/userdata/conmon.pid

[Install]
WantedBy=multi-user.target
```

Signed-off-by: Valentin Rothberg <rothberg@redhat.com>

### CODE REVIEW

**Comment by vrothberg** [comment]
@mheon @baude @rhatdan @edsantiago @TomSweeneyRedHat PTAL

Note, that the native client can now also generate files.

File: docs/podman-generate-systemd.1.md [hunk 1]
```diff
@@ -27,35 +30,74 @@ Set the systemd restart policy.  The restart-policy must be one of: "no", "on-su
...
 WantedBy=multi-user.target
 ```
 
-Create a systemd unit file for a container running nginx with an *always* restart policy and 1-second timeout.
+Create systemd unit files for a pod with a simple alpine container.
 ```
-$ sudo podman generate systemd --restart-policy=always -t 1 nginx
+$ podman pod create --name systemd-pod
+$ podman create --pod systemd-pod alpine top
+$ podman generate systemd --name systemd-pod
```
Comment (mheon) [inline docs/podman-generate-systemd.1.md]: So the command used here and below (the STDOUT one) are identical - missing the --files flag here?

File: docs/podman-generate-systemd.1.md [hunk 2]
```diff
@@ -27,35 +30,74 @@ Set the systemd restart policy.  The restart-policy must be one of: "no", "on-su
...
+# container-admiring_visvesvaraya.service
+[Unit]
+Description=admiring_visvesvaraya Podman Container
+Documentation=man:podman-generate-systemd(1)
+PartOf=pod-systemd-pod.service
+
+[Service]
+Restart=on-failure
+ExecStart=/home/valentin/go/src/github.com/containers/libpod/bin/podman start admiring_visvesvaraya
+ExecStop=/home/valentin/go/src/github.com/containers/libpod/bin/podman stop -t 10 admiring_visvesvaraya
```
Comment (mheon) [inline docs/podman-generate-systemd.1.md]: Should we be adding a timeout here? We're overriding whatever the user-specified timeout during creation was.

File: docs/podman-generate-systemd.1.md [hunk 3]
```diff
@@ -4,16 +4,19 @@
...
-to a file.  The options can be used to influence the results of the output as well.
+**podman generate systemd** will create a systemd unit file that can be used to control a container.
+By default, the command will print the content of the unit files on stdout.
 
 
 ## OPTIONS:
 
+**--files**, **-f**
+
+Generate files instead of printing on stdout. The files option is not supported on the remote client.
```
Comment (edsantiago) [inline docs/podman-generate-systemd.1.md]: Perhaps add something like `the files will be named random-string.service and will be created in the current working directory`?

File: test/e2e/generate_systemd_test.go [hunk 4]
```diff
@@ -71,4 +72,17 @@ var _ = Describe("Podman generate systemd", func() {
...
+		n.WaitWithDefaultTimeout()
+		Expect(n.ExitCode()).To(Equal(0))
+
+		n = podmanTest.Podman([]string{"create", "--pod", "foo", "alpine", "top"})
+		n.WaitWithDefaultTimeout()
+		Expect(n.ExitCode()).To(Equal(0))
+
+		session := podmanTest.Podman([]string{"generate", "systemd", "foo"})
+		session.WaitWithDefaultTimeout()
+		Expect(session.ExitCode()).To(Equal(0))
```
Comment (edsantiago) [inline test/e2e/generate_systemd_test.go]: y'know, this thing about "just check exit status but don't bother looking at output" troubles me greatly. Would it be possible to at least do something like look for 'Unit', 'ExecStart', 'WantedBy' in stdout?

File: docs/podman-generate-systemd.1.md (hunk 1 above)
Comment (vrothberg) [reply docs/podman-generate-systemd.1.md]: Good catch! Fixed.

File: docs/podman-generate-systemd.1.md (hunk 2 above)
Comment (vrothberg) [reply docs/podman-generate-systemd.1.md]: This behaviour pre-dates this PR. We're only overriding if the user sets the timeout explicitly via `generate systemd --timeout`.

File: docs/podman-generate-systemd.1.md (hunk 3 above)
Comment (vrothberg) [reply docs/podman-generate-systemd.1.md]: I like that, thanks! Done.

File: test/e2e/generate_systemd_test.go (hunk 4 above)
Comment (vrothberg) [reply test/e2e/generate_systemd_test.go]: I agree. I will add some tests now!

**Comment by edsantiago** [comment]
Minor nits:
* The '`Requires=` and `Before=` lines contain trailing whitespace. (I realize this is a preexisting condition; I'm just mentioning it in case it's easy to fix)
* The generated files do not include a trailing newline. This should be easy to fix.

File: test/e2e/generate_systemd_test.go (hunk 4 above)
Comment (vrothberg) [reply test/e2e/generate_systemd_test.go]: Added a commit to improve the tests. I'll add some unit tests on top as well.

**Comment by vrothberg** [comment]
> Minor nits:
> 
>     * The '`Requires=` and `Before=` lines contain trailing whitespace. (I realize this is a preexisting condition; I'm just mentioning it in case it's easy to fix)

That's actually introduced with this PR. I try to fix that ... go templates can be tricky. Thanks a lot for checking!

**Comment by vrothberg** [comment]
I will work on a few more unit tests and do some templating clean ups. In the meantime, I want to make sure that pod/container unit files make sense and that I didn't miss something ... @edsantiago @baude @mheon do they make sense to you? Ideally, could you play around and test the generated files?

**Comment by mheon** [comment]
We do have a systemd daemon available in "most" of the test VMs (exception is Podman-in-Podman). However, it's global state - interacting with it from threaded tests is inherently risky.

**Comment by vrothberg** [comment]
> We do have a systemd daemon available in "most" of the test VMs (exception is Podman-in-Podman). However, it's global state - interacting with it from threaded tests is inherently risky.

I share the though but I also feel that actually testing the files exceeds the PR.

**Comment by mheon** [comment]
Agree wholeheartedly. If we want to add testing, that sounds like a separate card and PR.

**Comment by edsantiago** [comment]
A few more thoughts:
* If the argument to `systemd generate` is a pod _name_, why not use that as the filename (and Description, ExecStart, etc) instead of the sha? Similarly if the container(s) have user-specified names.
* Did you consider writing the systemd files into their respective places (`/etc/systemd/system` or `$HOME/.config/systemd`)? I wouldn't necessarily recommend this, it makes me uncomfortable, I'm just wondering if you weighed the pros and cons.
* On autogenerated files, I like adding a comment at top showing a timestamp and 'autogenerated by COMMAND'.
* I can't actually get it to work as rootless: 
```
$ systemctl --user status container-59767e886456494857f68c14b4aa71a5b32917ebb42ebcf4ecd4a4c278c3b5a2.service | cat
Failed to start 59767e886456494857f68c14b4aa71a5b32917ebb42ebcf4ecd4a4c278c3b5a2 Podman Container
$ [ pod ]
pod-b7f9874412af55c8425187fc334aea8b565d49cb910c872f13a2fe0fef455c05.service: Start request repeated too quickly.
```
```

**Comment by vrothberg** [comment]
> A few more thoughts:
> 
>     * If the argument to `systemd generate` is a pod _name_, why not use that as the filename (and Description, ExecStart, etc) instead of the sha? Similarly if the container(s) have user-specified names.

That's possible when using the `--name` flag. Then we use the format `{container,pod}-name.service`.

>     * Did you consider writing the systemd files into their respective places (`/etc/systemd/system` or `$HOME/.config/systemd`)? I wouldn't necessarily recommend this, it makes me uncomfortable, I'm just wondering if you weighed the pros and cons.

Yes, I was thinking about writing the files to some configurable default but I prefer users to take care of that.

>     * On autogenerated files, I like adding a comment at top showing a timestamp and 'autogenerated by COMMAND'.

That's a great suggestion! Will do that.

>     * I can't actually get it to work as rootless:
> 
> 
> ```
> $ systemctl --user status container-59767e886456494857f68c14b4aa71a5b32917ebb42ebcf4ecd4a4c278c3b5a2.service | cat
> Failed to start 59767e886456494857f68c14b4aa71a5b32917ebb42ebcf4ecd4a4c278c3b5a2 Podman Container
> $ [ pod ]
> pod-b7f9874412af55c8425187fc334aea8b565d49cb910c872f13a2fe0fef455c05.service: Start request repeated too quickly.
> ```

Thanks a ton for checking. I've seen this error during development but didn't hit it after. Can you share the pod/containers you used?

**Comment by edsantiago** [comment]
Can you clarify "pod/containers you used"? I followed your trivial example 'foo' and 'alpine top'.

One thing I'm seeing is that the `conmon.pid` file is missing even though conmon is running and 'podman ps' lists the container running; I wonder if systemd is deleting it in certain circumstances? I'm continuing to investigate but wanted to share this early.

**Comment by vrothberg** [comment]
> Can you clarify "pod/containers you used"? I followed your trivial example 'foo' and 'alpine top'.

I tested the `foo` pod with a varying number of `alpine top` containers.

> One thing I'm seeing is that the `conmon.pid` file is missing even though conmon is running and 'podman ps' lists the container running; I wonder if systemd is deleting it in certain circumstances? I'm continuing to investigate but wanted to share this early.

That's very odd and screams for adding unit tests for the pod case.

**Comment by edsantiago** [comment]
@vrothberg try this for a container:
```
$ podman create --pod foo alpine sh -c 'echo Starting;while :;do date;sleep 1;done'
```
I'm seeing it restart every ten seconds (via `podman logs -f`); which looks suspiciously like the `-t 10` in the `ExecStop` but I'm not having luck figuring out why `stop` would be getting called.

**Comment by edsantiago** [comment]
I think this might be it. From `journalctl --user -xe`:
```
container-vibrant_cartwright.service: New main PID 24754 does not belong to service, and PID file is not owned by root. Refusing.
```
and, from `systemd.service(5)` under `PIDFile=`:
```
    The PID file does not need to be owned by a
    privileged user, but if it is owned by an unprivileged user additional safety
    restrictions are enforced: the file may not be a symlink to a file owned by a
    different user (neither directly nor indirectly), and the PID file must refer to
    a process already belonging to the service.
```
Could this be resolved by the requested `no-cgroup` option? And if so should I stop trying to test nonroot systemd, and in fact should you even restrict/disable the possibility of using `generate systemd` as nonroot?

**Comment by mheon** [comment]
`no-cgroup` would probably resolve this.

**Comment by vrothberg** [comment]
Looking into it now, thanks @edsantiago!

The headers now look like that:
```
# container-ced2fb140f0cfeb1c8fa21c2397b4d1ac1f48622265fdc7b4a759e5ff9b285fa.service
# autogenerated by Podman 1.5.1-dev
# Wed Aug 14 19:42:20 CEST 2019
```

**Comment by edsantiago** [comment]
vrothberg@irc> now I am hitting cases where the conmon.pid is missing for real.

I think that one of the cases I saw of missing-pidfile was because I was also manually playing with podman commands. systemd does not like this.

Thinking aloud: what if the generated unit files all used `podman --runroot /somewhere/else` (and possibly `--root` also)? This could isolate systemd-managed pods & containers.

(For the record I think the logistical hassles and complexity make this an unmanageable option; I'm throwing it out simply for the sake of possibility)

**Comment by vrothberg** [comment]
@edsantiago, if you have some cycles, I'd appreciate your thoughts and feedback on the latest state.

File: cmd/podman/generate_systemd.go [hunk 5]
```diff
@@ -40,7 +39,10 @@ func init() {
 	containerSystemdCommand.SetHelpTemplate(HelpTemplate())
 	containerSystemdCommand.SetUsageTemplate(UsageTemplate())
 	flags := containerSystemdCommand.Flags()
-	flags.BoolVarP(&containerSystemdCommand.Name, "name", "n", false, "use the container name instead of ID")
+	flags.BoolVarP(&containerSystemdCommand.Name, "name", "n", false, "use the container/pod name instead of ID (deprecated nop)")
```
Comment (vrothberg) [inline cmd/podman/generate_systemd.go]: Not a nop anymore :^) Had the impulse to do it. Will be fixed with the next revision.

File: docs/podman-generate-systemd.1.md [hunk 6]
```diff
@@ -27,35 +32,74 @@ Set the systemd restart policy.  The restart-policy must be one of: "no", "on-su
 "on-watchdog", "on-abort", or "always".  The default policy is *on-failure*.
 
 ## Examples
-Create a systemd unit file for a container running nginx:
 
+Create a systemd unit file for a container running nginx with an *always* restart policy and 1-second timeout on stdout.
```
Comment (vrothberg) [inline docs/podman-generate-systemd.1.md]: Examples must be updated with the latest state.

**Comment by edsantiago** [comment]
Generated files are still missing the final newline, could I insist on getting that fixed? 

Apart from that, it looks really good: it works almost exactly as expected. The only unexpected behavior is:
```
$ systemctl --user start pod-foo.service
$ podman ps
CONTAINER ID  IMAGE                            COMMAND               CREATED        STATUS           PORTS  NAMES
299f30b7b792  docker.io/library/alpine:latest  sh -c echo Start2...  2 minutes ago  Up 1 second ago         angry_shannon
67e686bacd06  docker.io/library/alpine:latest  sh -c echo Start1...  2 minutes ago  Up 1 second ago         quizzical_dewdney

$ podman stop  67e
67e686bacd0660d0f152737694bef4abf4dcf3cfa357988eda3b6cd167a52fe1

$ systemctl --user status container-quizzical_dewdney.service
● container-quizzical_dewdney.service - Podman container-quizzical_dewdney.service
   Loaded: loaded (/home/esm/.config/systemd/user/container-quizzical_dewdney.service; disabled; vendor preset: enabled)
   Active: inactive (dead) since Tue 2019-08-20 11:32:44 MDT; 37s ago
     Docs: man:podman-generate-systemd(1)
  Process: 2136 ExecStart=/home/esm/src/atomic/2018-02.podman/libpod/bin/podman start quizzical_dewdney (code=exited, status=0/SUCCESS)
  Process: 2402 ExecStop=/home/esm/src/atomic/2018-02.podman/libpod/bin/podman stop -t 10 quizzical_dewdney (code=exited, status=0/SUCCESS)
 Main PID: 2194 (code=exited, status=0/SUCCESS)

Aug 20 11:32:24 f.edsantiago.com systemd[1319]: Starting Podman container-quizzical_dewdney.service...
Aug 20 11:32:24 f.edsantiago.com podman[2136]: 2019-08-20 11:32:24.92576536 -0600 MDT m=+0.260550371 container init 67e686bacd0660d0f152737694bef4abf4dcf3cfa35>
Aug 20 11:32:24 f.edsantiago.com podman[2136]: 2019-08-20 11:32:24.934971566 -0600 MDT m=+0.269756578 container start 67e686bacd0660d0f152737694bef4abf4dcf3cfa>
Aug 20 11:32:24 f.edsantiago.com podman[2136]: quizzical_dewdney
Aug 20 11:32:24 f.edsantiago.com systemd[1319]: Started Podman container-quizzical_dewdney.service.
Aug 20 11:32:44 f.edsantiago.com podman[2194]: 2019-08-20 11:32:44.125152856 -0600 MDT m=+0.158152884 container cleanup 67e686bacd0660d0f152737694bef4abf4dcf3c>
Aug 20 11:32:44 f.edsantiago.com podman[2402]: 67e686bacd0660d0f152737694bef4abf4dcf3cfa357988eda3b6cd167a52fe1
Aug 20 11:32:44 f.edsantiago.com systemd[1319]: container-quizzical_dewdney.service: Succeeded.
```
That is: I started the service, stopped one container (using podman, behind systemd's back). I was expecting that container to restart. Is my expectation wrong?

Really nice work. That was a good idea with the infra container.

**Comment by baude** [comment]
@rhatdan you happy with this one?

**Comment by vrothberg** [comment]
> Generated files are still missing the final newline, could I insist on getting that fixed?

Yes, please insist :) I missed that one.

> [...]
> That is: I started the service, stopped one container (using podman, behind systemd's back). I was expecting that container to restart. Is my expectation wrong?

Your expectation is correct. In order for this to work, we need an updated conmon with https://github.com/containers/conmon/commit/74da3fb06d4444aca4724bddc8e61729600a4d81. I noticed the same issue and we tracked it down to old conmons not returning the exit code of the container which ultimately lets systemd believe that everything's in order.

> Really nice work. That was a good idea with the infra container.

Thanks! I am very happy that `podman pod ps` etc. continues working although not being started via `podman pod start`.

**Comment by vrothberg** [comment]
I am happy with the current state. As discussed, the support for the remote client has been dropped entirely.

@edsantiago @baude @mheon @rhatdan ready for final review.

**Comment by rhatdan** [comment]
@edsantiago You expectations are incorrect for now.  We had talked a while back about adding the feature to a pod to specify what happens when a container exits.  There should be three possible configurations.
1.  Nothing.   The pod and all other containers continue running
2.  Restart.  The container is automatically restarted
3.  Shutdown.  The pod and all other containers are stopped.

I would like to see this implemented.   Not sure if this needs to be tied to systemd.  I think making this a basic feature of pods would be nice.

File: docs/podman-generate-systemd.1.md [hunk 7]
```diff
@@ -28,41 +31,66 @@ Set the systemd restart policy.  The restart-policy must be one of: "no", "on-su
...
-Description=c21da63c4783be2ac2cd3487ef8d2ec15ee2a28f63dd8f145e3b05607f31cffc Podman Container
+Description=Podman container-de1e3223b1b888bc02d0962dd6cb5855eb00734061013ffdd3479d225abacdc6.service
+Documentation=man:podman-generate-systemd(1)
+
 [Service]
-Restart=on-failure
-ExecStart=/usr/bin/podman start c21da63c4783be2ac2cd3487ef8d2ec15ee2a28f63dd8f145e3b05607f31cffc
-ExecStop=/usr/bin/podman stop -t 10 c21da63c4783be2ac2cd3487ef8d2ec15ee2a28f63dd8f145e3b05607f31cffc
+Restart=always
+ExecStart=/home/valentin/go/src/github.com/containers/libpod/bin/podman start de1e3223b1b888bc02d0962dd6cb5855eb00734061013ffdd3479d225abacdc6
```
Comment (rhatdan) [inline docs/podman-generate-systemd.1.md]: I don't think you should have your name in the man page.  Should be 
/bin/podman start de1e3223b1b888bc02d0962dd6cb5855eb00734061013ffdd3479d225abacdc6

File: docs/podman-generate-systemd.1.md [hunk 8]
```diff
@@ -28,41 +31,66 @@ Set the systemd restart policy.  The restart-policy must be one of: "no", "on-su
...
+ExecStop=/home/valentin/go/src/github.com/containers/libpod/bin/podman stop -t 1 de1e3223b1b888bc02d0962dd6cb5855eb00734061013ffdd3479d225abacdc6
```
Comment (rhatdan) [inline docs/podman-generate-systemd.1.md]: Ditto

File: docs/podman-generate-systemd.1.md [hunk 9]
```diff
@@ -28,41 +31,66 @@ Set the systemd restart policy.  The restart-policy must be one of: "no", "on-su
...
 Type=forking
-PIDFile=/var/run/containers/storage/overlay-containers/c21da63c4783be2ac2cd3487ef8d2ec15ee2a28f63dd8f145e3b05607f31cffc/userdata/conmon.pid
+PIDFile=/run/user/1000/overlay-containers/de1e3223b1b888bc02d0962dd6cb5855eb00734061013ffdd3479d225abacdc6/userdata/conmon.pid
+
 [Install]
 WantedBy=multi-user.target
 ```
 
-Create a systemd unit file for a container running nginx with an *always* restart policy and 1-second timeout.
+Create systemd unit files for a pod with two simple alpine container. Note that the container services cannot be started/stopped manually but only the main one of the pod which starts/stops the services in the correct order.
```
Comment (rhatdan) [inline docs/podman-generate-systemd.1.md]: nit:
Create systemd unit files for a pod with two simple alpine container**s**

File: docs/podman-generate-systemd.1.md [hunk 10]
```diff
@@ -28,41 +31,66 @@ Set the systemd restart policy.  The restart-policy must be one of: "no", "on-su
...
+Documentation=man:podman-generate-systemd(1)
+Requires=container-amazing_chandrasekhar.service container-jolly_shtern.service
+Before=container-amazing_chandrasekhar.service container-jolly_shtern.service
+
 [Service]
-Restart=always
-ExecStart=/usr/bin/podman start c21da63c4783be2ac2cd3487ef8d2ec15ee2a28f63dd8f145e3b05607f31cffc
-ExecStop=/usr/bin/podman stop -t 1 c21da63c4783be2ac2cd3487ef8d2ec15ee2a28f63dd8f145e3b05607f31cffc
+Restart=on-failure
+ExecStart=/home/valentin/go/src/github.com/containers/libpod/bin/podman start 77a818221650-infra
```
Comment (rhatdan) [inline docs/podman-generate-systemd.1.md]: /bin/podman

File: docs/podman-generate-systemd.1.md [hunk 11]
```diff
@@ -28,41 +31,66 @@ Set the systemd restart policy.  The restart-policy must be one of: "no", "on-su
...
+ExecStop=/home/valentin/go/src/github.com/containers/libpod/bin/podman stop -t 10 77a818221650-infra
```
Comment (rhatdan) [inline docs/podman-generate-systemd.1.md]: /bin/podman

**Comment by vrothberg** [comment]
@rhatdan, I think that's an orthogonal topic. In systemd-service terms, the service should be restarted when we do a `podman stop $ctr` since it should have a non-zero exit code, was stopped via a signal and the service is configured to restart on-failure.

1) and 2) are working already with the patched version of conmon.

> 3) Shutdown. The pod and all other containers are stopped.

This will be trickier to map into systemd terms but should be possible when we alter the dependency graph.

File: docs/podman-generate-systemd.1.md (hunk 7 above)
Comment (vrothberg) [reply docs/podman-generate-systemd.1.md]: Thanks for checking!

File: docs/podman-generate-systemd.1.md (hunk 9 above)
Comment (vrothberg) [reply docs/podman-generate-systemd.1.md]: Corrected, thanks!

**Comment by vrothberg** [comment]
Corrected the typos, thanks for checking @rhatdan!

File: cmd/podman/generate_systemd.go [hunk 12]
```diff
@@ -40,7 +39,10 @@ func init() {
 	containerSystemdCommand.SetHelpTemplate(HelpTemplate())
 	containerSystemdCommand.SetUsageTemplate(UsageTemplate())
 	flags := containerSystemdCommand.Flags()
-	flags.BoolVarP(&containerSystemdCommand.Name, "name", "n", false, "use the container name instead of ID")
+	flags.BoolVarP(&containerSystemdCommand.Name, "name", "n", false, "use the container/pod name instead of ID")
+	if !remoteclient {
+		flags.BoolVarP(&containerSystemdCommand.Files, "files", "f", false, "generate files instead of printing on stdout")
```
Comment (TomSweeneyRedHat) [inline cmd/podman/generate_systemd.go]: nit, and only a suggestion:  s/on stdout/to stdout/

File: docs/podman-generate-systemd.1.md [hunk 13]
```diff
@@ -27,41 +31,66 @@ Set the systemd restart policy.  The restart-policy must be one of: "no", "on-su
...
 Type=forking
-PIDFile=/var/run/containers/storage/overlay-containers/c21da63c4783be2ac2cd3487ef8d2ec15ee2a28f63dd8f145e3b05607f31cffc/userdata/conmon.pid
+PIDFile=/run/user/1000/overlay-containers/de1e3223b1b888bc02d0962dd6cb5855eb00734061013ffdd3479d225abacdc6/userdata/conmon.pid
+
 [Install]
 WantedBy=multi-user.target
 ```
 
-Create a systemd unit file for a container running nginx with an *always* restart policy and 1-second timeout.
+Create systemd unit files for a pod with two simple alpine containers. Note that the container services cannot be started/stopped manually but only the main one of the pod which starts/stops the services in the correct order.
```
Comment (TomSweeneyRedHat) [inline docs/podman-generate-systemd.1.md]: I'm not really sure what you're trying to say in this second sentence.

**Comment by vrothberg** [comment]
Thanks for checking, @TomSweeneyRedHat! I will fix with the next push but am waiting for more feedback to avoid wasting energy with redundant CI runs.

File: docs/podman-generate.1.md [hunk 14]
```diff
@@ -11,10 +11,11 @@ The generate command will create structured output (like YAML) based on a contai
...
-| -------  | --------------------------------------------------- | ---------------------------------------------------------------------------- |
-| kube     | [podman-generate-kube(1)](podman-generate-kube.1.md)| Generate Kubernetes YAML based on a pod or container.                        |
-| systemd  | [podman-generate-systemd(1)](podman-generate-systemd.1.md)| Generate a systemd unit file for a container.                        |
+| Command | Man Page                                                   | Description                                                                        |
+|---------|------------------------------------------------------------|------------------------------------------------------------------------------------|
+| kube    | [podman-generate-kube(1)](podman-generate-kube.1.md)       | Generate Kubernetes YAML based on a pod or container.                              |
+| systemd | [podman-generate-systemd(1)](podman-generate-systemd.1.md) | Generate a systemd unit file for a container. Not supported for the remote client. |
+
```
Comment (edsantiago) [inline docs/podman-generate.1.md]: As long as you're waiting before push: how about:
```diff
- Generate a systemd unit file for a container
+ Generate systemd unit file(s) for a container or pod
```

File: docs/podman-generate-systemd.1.md (hunk 13 above)
Comment (edsantiago) [reply docs/podman-generate-systemd.1.md]: I think I understand what you mean, but agree with @TomSweeneyRedHat - the wording is confusing. I don't have any great suggestions, but maybe:
```
Note that these container services cannot be started or stopped individually via systemctl;
they are managed by the pod service. You can still use "systemctl status" or journalctl
to examine them.
```
(Nope, still not happy with that either, but I hope it gives you ideas)

**Comment by edsantiago** [comment]
Documentation nits only; all else LGTM. Thanks for your perseverance.

File: cmd/podman/generate_systemd.go (hunk 12 above)
Comment (vrothberg) [reply cmd/podman/generate_systemd.go]: Good one, thanks!

File: docs/podman-generate-systemd.1.md (hunk 13 above)
Comment (vrothberg) [reply docs/podman-generate-systemd.1.md]: Thanks folks! I like Ed's text and will copy it :)

File: docs/podman-generate.1.md (hunk 14 above)
Comment (vrothberg) [reply docs/podman-generate.1.md]: Good catch, thanks!

**Comment by vrothberg** [comment]
Thanks for the review, @edsantiago and @TomSweeneyRedHat! I just updated the PR.

File: docs/podman-generate-systemd.1.md [hunk 15]
```diff
@@ -4,16 +4,20 @@
...
 ## SYNOPSIS
-**podman generate systemd** [*options*] *container*
+**podman generate systemd** [*options*] *container|pod*
 
 ## DESCRIPTION
-**podman generate systemd** will create a Systemd unit file that can be used to control a container.  The
-command will dynamically create the unit file and output it to stdout where it can be piped by the user
-to a file.  The options can be used to influence the results of the output as well.
+**podman generate systemd** will create a systemd unit file that can be used to control a container or pod.
+By default, the command will print the content of the unit files on stdout.
```
Comment (TomSweeneyRedHat) [inline docs/podman-generate-systemd.1.md]: sorry, missed this earlier s/on stdout/to stdout/

File: docs/podman-generate-systemd.1.md [hunk 16]
```diff
@@ -27,41 +31,66 @@ Set the systemd restart policy.  The restart-policy must be one of: "no", "on-su
...
 Type=forking
-PIDFile=/var/run/containers/storage/overlay-containers/c21da63c4783be2ac2cd3487ef8d2ec15ee2a28f63dd8f145e3b05607f31cffc/userdata/conmon.pid
+PIDFile=/run/user/1000/overlay-containers/de1e3223b1b888bc02d0962dd6cb5855eb00734061013ffdd3479d225abacdc6/userdata/conmon.pid
+
 [Install]
 WantedBy=multi-user.target
 ```
 
-Create a systemd unit file for a container running nginx with an *always* restart policy and 1-second timeout.
+Create systemd unit files for a pod with two simple alpine containers. Note that these container services cannot be started or stopped individually via `systemctl`; they are managed by the pod service. You can still use `systemctl status` or journalctl to examine them.
```
Comment (TomSweeneyRedHat) [inline docs/podman-generate-systemd.1.md]: :+1:

File: docs/podman-generate-systemd.1.md (hunk 15 above)
Comment (vrothberg) [reply docs/podman-generate-systemd.1.md]: 44 seconds until push :^)

**Comment by TomSweeneyRedHat** [comment]
LGTM, nice work @vrothberg

**Comment by baude** [comment]
@rhatdan drop the hold if you happy with the pr now

**Pull request merged**